File: new-algorithm/src/javaalgo/stack/boj9012.java
package javaalgo.stack;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class boj9012 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());

        while (T-- > 0) {
            sb.append(solution(br.readLine())).append('\n');
        }

        System.out.println(sb);
    }

    static String solution(String s) {
        Stack<Character> stack = new Stack<>();
        char[] arr = s.toCharArray();

        for (char c : arr) {
            if (c == '(') {
                stack.push(c);
            } else {
                if (stack.size() == 0) stack.push(c);
                if (stack.peek() == '(') {
                    stack.pop();
                }
            }
        }

        if (stack.size() == 0) return "YES";
        else return "NO";
    }
}

File: new-algorithm/src/kotlinalgo/stack/boj9012.kt
package kotlinalgo.stack

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.Stack

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    var T = Integer.parseInt(readLine())
    val sb = StringBuilder()

    while (T-- != 0) {
        sb.append(findVPS(readLine())).append('\n')
    }

    println(sb.toString())

}

fun findVPS(s: String): String {
    val arr = s.toCharArray()
    val stack = Stack<Char>()

    for (c in arr) {
        when (c) {
            '(' -> stack.push(c)
            else -> when {
                stack.size == 0 -> stack.push(c)
                stack.peek() == '(' -> stack.pop()
            }
        }
    }

    return if (stack.size == 0) "YES" else "NO"

}